### DISCUSSION TITLE
script record sound and slide transition times

### DISCUSSION THREAD

**Discussion by ousia** [Cimbali/pympress#221] [Q&A] [answered]
@Cimbali,

I wonder whether it would be possible to have in a Python3 script that records sound while Pympress displays a presentation, writing slide transition times to a text file.

The same way that a presentation is started by pressing space, this would start sound recording and mark the starting time for the transition times.

After that, each slide transition time should be written to a text file (page number isn’t required in the text file). Once the last slide is reached, the next pressing of the space key should stop sound recording (and write the last slide transition time [which actually contains its duration]).

Many thanks for your help.

**Comment by Cimbali** [comment, marked as answer]
I think a key element to write that script here would be to use DBus messages that pympress emits on play/pause and slide transitions. That’s one of the nice side effects of using the Gtk.Application (since v1.6.0).

Below some excerpts from a `dbus-monitor` output that could be useful.

On document load, page set to 1 and timer paused:
```
signal time=1636062833.380317 sender=:1.962 -> destination=(null destination) serial=117 path=/io/github/pympress; interface=org.gtk.Actions; member=Changed
   array [
   ]
   array [
   ]
   array [
      dict entry(
         string "pause-timer"
         variant             boolean true
      )
      dict entry(
         string "page"
         variant             int64 1
      )
   ]
   array [
   ]
```
Un-pause timer:
```
signal time=1636062882.262072 sender=:1.962 -> destination=(null destination) serial=118 path=/io/github/pympress; interface=org.gtk.Actions; member=Changed
   array [
   ]
   array [
   ]
   array [
      dict entry(
         string "pause-timer"
         variant             boolean false
      )
   ]
   array [
   ]
```
page change:
```
signal time=1636062894.224972 sender=:1.962 -> destination=(null destination) serial=119 path=/io/github/pympress; interface=org.gtk.Actions; member=Changed
   array [
   ]
   array [
   ]
   array [
      dict entry(
         string "page"
         variant             int64 2
      )
   ]
   array [
   ]
```

Following the [pretty good dbus tutorial here](https://dbus.freedesktop.org/doc/dbus-python/index.html), I managed to write the following (rather simple) code:

```python
#!/usr/bin/env python3

import dbus
from dbus.mainloop.glib import DBusGMainLoop

import time
import gi
from gi.repository import GLib

class PympressTimer:
    def __init__(self):
        self.paused_at = 0
        self.start = None
        self.page = 0

    def handle(self, *messages):
        timestamp = time.time()
        for msg in messages:
            if not isinstance(msg, dbus.Dictionary) or msg.signature != 'sv':
                continue
            if 'pause-timer' in msg:
                if msg['pause-timer']:
                    self.paused_at = timestamp
                else:
                    self.start = timestamp - self.paused_at
                    self.paused_at = None
            if 'page' in msg:
                self.page = msg['page']
            if self.start is not None:
                print(f'{timestamp - self.start:4.2f} page', self.page)

    def run(self):
        DBusGMainLoop(set_as_default=True)
        bus = dbus.SessionBus()
        bus.add_signal_receiver(self.handle, 'Changed', 'org.gtk.Actions', 'io.github.pympress')

        loop = GLib.MainLoop()
        loop.run()

if __name__ == '__main__':
    PympressTimer().run()
```
This tracks whenever page transitions are made and only counts up the time when the pympress timer is un-paused

**Reply by ousia** [reply]
Many thanks for the basic script, @Cimbali.

I have a basic question too: how can I run the script using the Windows binaries?

I mean, I don’t have Python installed on the Windows machine and I don’t know how to invoke it.

Many thanks for your help again.

**Reply by Cimbali** [reply]
I think you’d need to package it with cx_Freeze from a machine that has python. The easiest is to install [msys2](https://www.msys2.org/) I believe, be it on the build machine or the final machine. From there install python and everything you need.

This can be done built on github servers, as we do from the build github action for example:

https://github.com/Cimbali/pympress/blob/b6e2091f581c05ad3176dd76717c085d32cacf5f/.github/workflows/draft_release.yml#L77-L146

**Comment by ousia** [comment]
I don’t know whether I should open a bug report, but for some strange reason `dbus-monitor` is only able to display pause/unpause actions, but not page changes. So the script only prints the pause and unpause times.

I’m on Fedora 35 and I cannot imagine which dependency I might be missing. I’m using version 1.7b1.

Is this a bug or what am I missing?

Many thanks for your help.

**Reply by Cimbali** [reply]
I added the page number action after the 1.7.0b1, so you'll need latest master

**Reply by ousia** [reply]
Sorry, I didn’t know that. I’m going to wait until the next beta release.
